lib/Reader/abisignature.cpp

```diff
@@ -96,18 +96,33 @@ ABISignature::ABISignature(const ReaderCallSignature &Signature, GenIR &Reader,
 Value *ABISignature::coerce(GenIR &Reader, Type *TheType, Value *TheValue) {
   assert(!TheType->isVoidTy());
 
-  Type *ValueType = TheValue->getType();
-
-  if (TheType == ValueType) {
-    return TheValue;
+  bool PointerRepresentsStruct = Reader.doesValueRepresentStruct(TheValue);
+  Type *ValueType = nullptr;
+  Value *ValuePtr = nullptr;
+  if (PointerRepresentsStruct) {
+    ValueType = TheValue->getType()->getPointerElementType();
+    if ((TheType == ValueType) && TheType->isStructTy()) {
```

Comment:
I was wrong, I got this case confused with something else. I took your suggestion.